app/models/bulk_upload.rb

```diff
@@ -11,11 +12,56 @@ class BulkUpload < ApplicationRecord
 
   after_initialize :generate_identifier, unless: :identifier
 
+  scope :search_by_filename, ->(filename) { where("filename ILIKE ?", "%#{filename}%") }
+  scope :search_by_user_name, ->(name) { where(user_id: User.where("name ILIKE ?", "%#{name}%").select(:id)) }
+  scope :search_by_user_email, ->(email) { where(user_id: User.where("email ILIKE ?", "%#{email}%").select(:id)) }
+  scope :search_by_organisation_name, ->(name) { where(organisation_id: Organisation.where("name ILIKE ?", "%#{name}%").select(:id)) }
+
+  scope :search_by, lambda { |param|
+    search_by_filename(param)
+      .or(search_by_user_name(param))
+      .or(search_by_user_email(param))
+      .or(search_by_organisation_name(param))
+  }
+
+  scope :filter_by_id, ->(id) { where(id:) }
+  scope :filter_by_years, lambda { |years, _user = nil|
+    first_year = years.shift
+    query = where(year: first_year)
+    years.each { |year| query = query.or(where(year:)) }
+    query.all
+  }
+  scope :filter_by_uploaded_by, ->(user_id, _user = nil) { where(user_id:) }
+  scope :filter_by_user_text_search, ->(param, _user = nil) { where(user_id: User.search_by(param).select(:id)) }
+  scope :filter_by_user, ->(user_id, _user = nil) { user_id.present? ? where(user_id:) : all }
+  scope :filter_by_uploading_organisation, ->(organisation_id, _user = nil) { where(organisation_id:) }
+
   def completed?
     incomplete_logs = logs.where.not(status: "completed")
     !incomplete_logs.exists?
   end
 
+  def status
+    return :processing if processing
+    return :blank_template if failed == "blank_template"
+    return :wrong_template if failed == "wrong_template"
+
+    if logs.visible.exists?
+      return :errors_fixed_in_service if completed? && bulk_upload_errors.any?
+      return :logs_uploaded_with_errors if bulk_upload_errors.any?
+    end
+
+    if bulk_upload_errors.any? { |error| error.category == "setup" }
```

Comment:
would it make sense for these to be scopes? `bulk_upload_errors.setup.exist?`

Reply:
Ye good idea